pkg/cmd/roachtest/tests/lease.go

```diff
@@ -0,0 +1,187 @@
+//
+// The cluster layout is as follows:
+//
+// n1-n3: System ranges and SQL gateways.
+// n4-n6: Workload ranges.
+// n7:    Workload runner.
+//
+// The test runs a kv50 workload with batch size 1, using 256 concurrent
+// workers directed at n1-n3. n4-n6 are killed and restarted in order, with 30
+// seconds between each operation, for 3 cycles totaling 9 crashes.
+func runAcquireLeaseCrashNonSystem(ctx context.Context, t test.Test, c cluster.Cluster) {
+	c.Put(ctx, t.Cockroach(), "./cockroach")
+	c.Put(ctx, t.DeprecatedWorkload(), "./workload")
+
+	// Create cluster with one rack locality per node and configure it.
+	require.Equal(t, 7, c.Spec().NodeCount)
+
+	opts := option.DefaultStartOpts()
+	settings := install.MakeClusterSettings(install.NumRacksOption(c.Spec().NodeCount))
+	c.Start(ctx, t.L(), opts, settings, c.Range(1, 6))
+
+	conn := c.Conn(ctx, t.L(), 1)
+
+	t.Status("configuring cluster")
+	_, err := conn.ExecContext(ctx, `SET CLUSTER SETTING kv.range_split.by_load_enabled = 'false'`)
+	require.NoError(t, err)
+
+	// Constrain all existing zone configs to n1-n3.
+	conn2 := c.Conn(ctx, t.L(), 1)
+	rows, err := conn.QueryContext(ctx, `SELECT target FROM [SHOW ALL ZONE CONFIGURATIONS]`)
+	require.NoError(t, err)
+	for rows.Next() {
+		var target string
+		require.NoError(t, rows.Scan(&target))
+		_, err = conn2.ExecContext(ctx, fmt.Sprintf(
+			`ALTER %s CONFIGURE ZONE USING num_replicas = 3, constraints = '[-rack=3, -rack=4, -rack=5]'`,
+			target))
+		require.NoError(t, err)
+	}
+	require.NoError(t, rows.Err())
+	rows.Close()
+	require.NoError(t, conn2.Close())
+
+	// Wait for upreplication.
+	require.NoError(t, WaitFor3XReplication(ctx, t, conn))
+
+	// Create the kv database, constrained to n4-n6. Despite the zone config, the
+	// ranges will initially be distributed across all cluster nodes.
+	t.Status("creating workload database")
+	_, err = conn.ExecContext(ctx, `CREATE DATABASE kv`)
+	require.NoError(t, err)
+	_, err = conn.ExecContext(ctx, `ALTER DATABASE kv CONFIGURE ZONE USING `+
+		`num_replicas = 3, constraints = '[-rack=0, -rack=1, -rack=2]'`)
+	require.NoError(t, err)
+	c.Run(ctx, c.Node(7), `./workload init kv --splits 1000 {pgurl:1}`)
+
+	// The replicate queue takes forever to move the kv ranges from n1-n3 to
+	// n4-n6, so we do it ourselves.
+	var count int
+	for _, source := range []int{1, 2, 3} {
+		for {
+			require.NoError(t, conn.QueryRowContext(ctx, `SELECT count(*) FROM crdb_internal.ranges `+
+				`WHERE database_name = 'kv' AND $1::int = ANY(replicas)`, source).Scan(&count))
+			if count == 0 {
+				break
+			}
+			t.Status(fmt.Sprintf("moving %d kv ranges off of n%d", count, source))
+			for _, target := range []int{4, 5, 6} {
+				_, err = conn.ExecContext(ctx, `ALTER RANGE RELOCATE FROM $1::int TO $2::int FOR `+
+					`SELECT range_id FROM crdb_internal.ranges WHERE database_name = 'kv' AND $1::int = ANY(replicas)`,
+					source, target)
+				require.NoError(t, err)
```

Comment:
Isn't this just gonna flake? Or is `ALTER RANGE RELOCATE` guaranteed to retry all errors away internally?

Reply:
It outputs errors as strings in the result table.